### PR TITLE
[SFT-1390]: Ability to round result values on Calculation fields

### PR DESCRIPTION
**Pull request by thejahid**

### CODE REVIEW

File: packages/plugin/src/Bundles/Fields/Validation/CalculationFieldValidation.php [hunk 1]
```diff
@@ -58,8 +59,11 @@ public function validate(ValidateEvent $event): void
 
         try {
             $result = $expressionLanguage->evaluate($calculationLogic, $variablesWithValue);
+            if ($decimalCount) {
+                $result = number_format($result, $decimalCount);
+            }
 
-            if ($canRender && $valueOrdination != $result) {
+            if ($valueOrdination != $result) {
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Fields/Validation/CalculationFieldValidation.php:66]: I doubt the error is going to be rendered on a hidden field.

If `$canRender` is false, the error should be appended to the form errors list `$form->addError();`

File: packages/plugin/src/Form/Form.php [hunk 2]
```diff
@@ -670,11 +671,20 @@ public function renderTag(?array $renderProperties = null): Markup
 
         $output .= '<form'.$this->getAttributes().$this->getAttributes()->getForm().'>'.\PHP_EOL;
 
-        $hiddenFields = $this->getCurrentPage()->getFields()->getList(HiddenField::class);
+        $fields = $this->getCurrentPage()->getFields();
+        $hiddenFields = $fields->getList(HiddenField::class);
+        $calculationFields = $fields->getList(CalculationField::class);
+
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Form/Form.php]: As we discussed:
* the logic for hidden fields should be extracted out from here and into its own feature bundle
* the logic for rendering hidden calculation fields should be extracted into its own feature bundle as well

So in the end there would be two new feature bundles.

File: packages/plugin/src/templates/submissions/edit.twig [hunk 3]
```diff
@@ -16,60 +16,54 @@
...
-                <li>
-                    <a class="formsubmit" data-redirect="{{ continueEditingUrl|hash }}">
-                        {{ "Save and continue editing"|t('app') }}
-                        <span class="shortcut">⌘S</span>
-                    </a>
-                </li>
-            </ul>
-        </div>
-    </div>
+	<div class="btngroup submit">
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/templates/submissions/edit.twig:19]: These file changes seem to be in error, is there anything new here, apart from formatting changes?

The formatting changes don't look better...

File: packages/scripts/src/lib/plugin/handlers/fields/calculations.ts [hunk 4]
```diff
@@ -48,33 +50,31 @@ class Calculation implements FreeformHandler {
       const doCalculations = () => {
         const allVariablesHaveValues = Object.values(variables).every((value) => value !== '');
 
-        if (!allVariablesHaveValues) {
-          return;
-        }
+        if (!allVariablesHaveValues) return;
```
Comment (gustavs-gutmanis) [inline packages/scripts/src/lib/plugin/handlers/fields/calculations.ts]: Lets stick to our code style and refrain from using one line conditional rules.

```suggestion
        if (!allVariablesHaveValues) {
          return;
        }
```

File: packages/scripts/src/lib/plugin/handlers/fields/calculations.ts [hunk 5]
```diff
@@ -48,33 +50,31 @@ class Calculation implements FreeformHandler {
...
 
         const result = this.expressionLanguage.evaluate(calculationsLogic, variables);
+        if (!(picker instanceof HTMLInputElement)) return;
```
Comment (gustavs-gutmanis) [inline packages/scripts/src/lib/plugin/handlers/fields/calculations.ts]: Same here

File: packages/plugin/src/Bundles/Fields/Validation/CalculationFieldValidation.php [hunk 6]
```diff
@@ -64,7 +64,15 @@ public function validate(ValidateEvent $event): void
...
-                $field->addError(Freeform::t('Incorrectly calculated value'));
+                $errorMessage = Freeform::t('Incorrectly calculated value');
+
+                if (!$canRender) {
+                    $form->addError($errorMessage);
+
+                    return;
+                }
+
+                $field->addError($errorMessage);
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Fields/Validation/CalculationFieldValidation.php]: If possible, lets strive to use positive evaluations first, they're a bit easier to read when going through the code fast.



```suggestion
                $errorMessage = Freeform::t('Incorrectly calculated value');
                if ($canRender) {
                    $field->addError($errorMessage);
                } else {
                    $form->addError($errorMessage);
                }
```

File: packages/plugin/src/Bundles/Form/HiddenInputs/CalculationFormFields.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Solspace\Freeform\Bundles\Form\HiddenInputs;
+
+use Solspace\Freeform\Events\Forms\RenderTagEvent;
+use Solspace\Freeform\Fields\Implementations\Pro\CalculationField;
+use Solspace\Freeform\Form\Form;
+use Solspace\Freeform\Library\Bundles\FeatureBundle;
+use yii\base\Event;
+
+class CalculationFormFields extends FeatureBundle
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Form/HiddenInputs/CalculationFormFields.php]: As I mentioned in Slack - lets move this into:
`packages/plugin/src/Bundles/Fields/Implementations/ConditionalField/ConditionalFieldBundle.php`

**Review by gustavs-gutmanis — DISMISSED** [review]

**Review by kjmartens — APPROVED** [review]

**Pull request merged**